src/algebra/Module/basic.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2019 Robert A. Spencer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert A. Spencer
+-/
+import algebra.module
+import algebra.punit_instances
+import category_theory.types
+import linear_algebra.basic
+open category_theory
+
+universe u
+
+variables (R : Type u) [ring R]
+
+/-- An example of a module for any ring is the zero module -/
+instance module.zero_module : module R punit := module.of_core $
+  by refine {
+    smul := λ _ _, punit.star,
+    .. punit.comm_ring, ..};
+    intros; exact subsingleton.elim _ _
+
+/-- The category of R-modules and their morphisms. -/
+structure Module :=
+  (carrier : Type)
+  (prop_add_comm_group : add_comm_group carrier)
+  (prop_module : module R carrier)
+
+namespace Module
+  instance : has_coe_to_sort (Module R) :=
+    { S := Type, coe := Module.carrier}
+
+  instance add_comm_group (M : Module R) : add_comm_group M := M.prop_add_comm_group
+  instance R_module (M : Module R) : module R M := M.prop_module
+
+  def of (X : Type) [h₁ : add_comm_group X] [h₂ : module R X] : Module R := ⟨ X , h₁ , h₂⟩
+
+  instance : has_zero (Module R) := ⟨ of R punit ⟩
+
+  variables (M N U : Module R)
+
+  instance : category (Module R) := {
+    hom := λ M N, M →ₗ[R] N,
+    id := λ M, 1 ,
+    comp := λ A B C f g, g.comp f ,
+  }
+
+  @[simp] lemma module_id : linear_map.to_fun (𝟙 M) = id := rfl
```

Comment:
I wonder if `id_apply (m : M) : (𝟙 M) m = m` is a more useful form of this simp-lemma.
Also, people seem to be a bit divided about whether or not to use the `.to_fun` projection, or coercions to functions... I'm personally in favour of coercions, but Scott seems to prefer the projections.

Reply:
I agree `id_apply` is more likely to be safe.

(Regarding coercions --- I just get really frustrated when they don't fire automatically, and I find myself jumping through extra hoops to use them. Generally, my attitude is that we're sufficiently far from "natural english"/"easily readable to a general mathematician" that there's little point improving notation unless it's very smooth.)

Reply:
Hmm.  I don't really want to pick a side, because I don't understand the merits of each.  For this PR though I think I'll go with coercion, since I don't want to have to remember that to get a function from a morphism requires `linear_map.to_fun`.  Just casting it seems easier.